File: android/app/src/test/java/com/loginscreen/LoginViewModelTest.kt
package com.loginscreen

import kotlinx.coroutines.flow.first
import kotlinx.coroutines.test.StandardTestDispatcher
import kotlinx.coroutines.test.advanceUntilIdle
import kotlinx.coroutines.test.runTest
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.test.setMain
import kotlinx.coroutines.test.resetMain
import org.junit.Assert.*
import org.junit.Before
import org.junit.After
import org.junit.Test
import org.mockito.kotlin.*

class LoginViewModelTest {

    private lateinit var authRepository: AuthRepository
    private lateinit var networkMonitor: NetworkMonitor
    private lateinit var tokenStorage: TokenStorage
    private lateinit var viewModel: LoginViewModel
    private val testDispatcher = StandardTestDispatcher()

    @Before
    fun setup() {
        Dispatchers.setMain(testDispatcher)
        authRepository = mock()
        networkMonitor = mock()
        tokenStorage = mock()
    }

    @After
    fun tearDown() {
        Dispatchers.resetMain()
    }

    @Test
    fun `test validation enables disables button`() = runTest(testDispatcher) {
        // Given: Network is online
        whenever(networkMonitor.isOnline).thenReturn(
            kotlinx.coroutines.flow.flowOf(true)
        )

        viewModel = LoginViewModel(authRepository, networkMonitor, tokenStorage)

        // When: Username and password are empty
        var state = viewModel.uiState.first()
        assertFalse("Button should be disabled when fields are empty", state.isButtonEnabled)

        // When: Only username is filled
        viewModel.updateUsername("user")
        advanceUntilIdle()
        state = viewModel.uiState.first()
        assertFalse("Button should be disabled when password is empty", state.isButtonEnabled)

        // When: Only password is filled
        viewModel.updatePassword("")
        viewModel.updateUsername("")
        viewModel.updatePassword("password")
        advanceUntilIdle()
        state = viewModel.uiState.first()
        assertFalse("Button should be disabled when username is empty", state.isButtonEnabled)

        // When: Both fields are filled
        viewModel.updateUsername("user")
        viewModel.updatePassword("password")
        advanceUntilIdle()
        state = viewModel.uiState.first()
        assertTrue("Button should be enabled when both fields are filled", state.isButtonEnabled)
    }

    @Test
    fun `test success navigation event`() = runTest(testDispatcher) {
        // Given: Network is online and valid credentials
        whenever(networkMonitor.isOnline).thenReturn(
            kotlinx.coroutines.flow.flowOf(true)
        )
        whenever(authRepository.login("user", "password")).thenReturn(
            Result.success("token123")
        )

        viewModel = LoginViewModel(authRepository, networkMonitor, tokenStorage)

        // When: Valid credentials are entered and login is called
        viewModel.updateUsername("user")
        viewModel.updatePassword("password")
        advanceUntilIdle()

        viewModel.login()
        advanceUntilIdle()

        // Then: Navigation event should be triggered
        val state = viewModel.uiState.first()
        assertNotNull("Navigation event should be set", state.navigationEvent)
        assertTrue("Navigation event should be NavigateToHome", 
            state.navigationEvent is NavigationEvent.NavigateToHome)
        assertNull("Error message should be null on success", state.errorMessage)
        assertEquals("Failure count should be reset", 0, state.failureCount)
    }

    @Test
    fun `test error increments failure count`() = runTest(testDispatcher) {
        // Given: Network is online and invalid credentials
        whenever(networkMonitor.isOnline).thenReturn(
            kotlinx.coroutines.flow.flowOf(true)
        )
        whenever(authRepository.login(any(), any())).thenReturn(
            Result.failure(Exception("Invalid credentials"))
        )

        viewModel = LoginViewModel(authRepository, networkMonitor, tokenStorage)

        viewModel.updateUsername("wrong")
        viewModel.updatePassword("wrong")
        advanceUntilIdle()

        // When: Login fails
        viewModel.login()
        advanceUntilIdle()

        // Then: Failure count should be incremented
        var state = viewModel.uiState.first()
        assertEquals("Failure count should be 1", 1, state.failureCount)
        assertNotNull("Error message should be set", state.errorMessage)
        assertFalse("Should not be locked out after 1 failure", state.isLockedOut)

        // When: Login fails again
        viewModel.login()
        advanceUntilIdle()
        state = viewModel.uiState.first()
        assertEquals("Failure count should be 2", 2, state.failureCount)
        assertFalse("Should not be locked out after 2 failures", state.isLockedOut)
    }

    @Test
    fun `test lockout after 3 failures`() = runTest(testDispatcher) {
        // Given: Network is online and invalid credentials
        whenever(networkMonitor.isOnline).thenReturn(
            kotlinx.coroutines.flow.flowOf(true)
        )
        whenever(authRepository.login(any(), any())).thenReturn(
            Result.failure(Exception("Invalid credentials"))
        )

        viewModel = LoginViewModel(authRepository, networkMonitor, tokenStorage)

        viewModel.updateUsername("wrong")
        viewModel.updatePassword("wrong")
        advanceUntilIdle()

        // When: Login fails 3 times
        viewModel.login()
        advanceUntilIdle()
        var state = viewModel.uiState.first()
        assertEquals("Failure count should be 1", 1, state.failureCount)

        viewModel.login()
        advanceUntilIdle()
        state = viewModel.uiState.first()
        assertEquals("Failure count should be 2", 2, state.failureCount)

        viewModel.login()
        advanceUntilIdle()
        state = viewModel.uiState.first()

        // Then: Account should be locked out
        assertEquals("Failure count should be 3", 3, state.failureCount)
        assertTrue("Account should be locked out", state.isLockedOut)
        assertFalse("Button should be disabled when locked out", state.isButtonEnabled)
    }

    @Test
    fun `test offline shows message and no service call`() = runTest(testDispatcher) {
        // Given: Network is offline
        whenever(networkMonitor.isOnline).thenReturn(
            kotlinx.coroutines.flow.flowOf(false)
        )

        viewModel = LoginViewModel(authRepository, networkMonitor, tokenStorage)

        viewModel.updateUsername("user")
        viewModel.updatePassword("password")
        advanceUntilIdle()

        // When: User tries to login while offline
        val state = viewModel.uiState.first()
        
        // Then: Offline message should be shown
        assertTrue("Should show offline state", state.isOffline)
        assertFalse("Button should be disabled when offline", state.isButtonEnabled)

        // When: Login is attempted (should not call service)
        viewModel.login()
        advanceUntilIdle()

        // Then: AuthRepository should not be called
        verify(authRepository, never()).login(any(), any())
    }

    @Test
    fun `test remember me persists token`() = runTest(testDispatcher) {
        // Given: Network is online and valid credentials
        val testToken = "test_token_123"
        whenever(networkMonitor.isOnline).thenReturn(
            kotlinx.coroutines.flow.flowOf(true)
        )
        whenever(authRepository.login("user", "password")).thenReturn(
            Result.success(testToken)
        )

        viewModel = LoginViewModel(authRepository, networkMonitor, tokenStorage)

        viewModel.updateUsername("user")
        viewModel.updatePassword("password")
        viewModel.updateRememberMe(true)
        advanceUntilIdle()

        // When: Login succeeds with remember me enabled
        viewModel.login()
        advanceUntilIdle()

        // Then: Token should be saved
        verify(tokenStorage).saveToken(testToken)

        // When: Remember me is disabled
        viewModel.updateRememberMe(false)
        viewModel.updateUsername("user2")
        viewModel.updatePassword("password2")
        whenever(authRepository.login("user2", "password2")).thenReturn(
            Result.success("token2")
        )
        viewModel.login()
        advanceUntilIdle()

        // Then: Token should be cleared
        verify(tokenStorage).clearToken()
    }
}

File: ios/LoginScreen/LoginScreenTests/LoginViewModelTest.swift
import XCTest
import Combine

@MainActor
class LoginViewModelTest: XCTestCase {
    var authService: MockAuthService!
    var networkMonitor: MockNetworkMonitor!
    var tokenStorage: MockTokenStorage!
    var viewModel: LoginViewModel!
    var cancellables: Set<AnyCancellable>!
    
    override func setUp() {
        super.setUp()
        authService = MockAuthService()
        networkMonitor = MockNetworkMonitor()
        tokenStorage = MockTokenStorage()
        cancellables = Set<AnyCancellable>()
    }
    
    override func tearDown() {
        cancellables = nil
        viewModel = nil
        tokenStorage = nil
        networkMonitor = nil
        authService = nil
        super.tearDown()
    }
    
    func testValidationEnablesDisablesButton() async {
        // Given: Network is online
        networkMonitor.isOnlineSubject.send(true)
        
        viewModel = LoginViewModel(
            authService: authService,
            networkMonitor: networkMonitor,
            tokenStorage: tokenStorage
        )
        
        // Wait for initial state
        try? await Task.sleep(nanoseconds: 100_000_000) // 0.1 seconds
        
        // When: Username and password are empty
        var expectation = XCTestExpectation(description: "Button disabled when empty")
        viewModel.$state
            .map { $0.isButtonEnabled }
            .sink { isEnabled in
                if !isEnabled {
                    expectation.fulfill()
                }
            }
            .store(in: &cancellables)
        
        await fulfillment(of: [expectation], timeout: 1.0)
        XCTAssertFalse(viewModel.state.isButtonEnabled, "Button should be disabled when fields are empty")
        
        // When: Only username is filled
        viewModel.updateUsername("user")
        try? await Task.sleep(nanoseconds: 100_000_000)
        XCTAssertFalse(viewModel.state.isButtonEnabled, "Button should be disabled when password is empty")
        
        // When: Only password is filled
        viewModel.updateUsername("")
        viewModel.updatePassword("password")
        try? await Task.sleep(nanoseconds: 100_000_000)
        XCTAssertFalse(viewModel.state.isButtonEnabled, "Button should be disabled when username is empty")
        
        // When: Both fields are filled
        viewModel.updateUsername("user")
        viewModel.updatePassword("password")
        try? await Task.sleep(nanoseconds: 100_000_000)
        XCTAssertTrue(viewModel.state.isButtonEnabled, "Button should be enabled when both fields are filled")
    }
    
    func testSuccessNavigationEvent() async {
        // Given: Network is online and valid credentials
        networkMonitor.isOnlineSubject.send(true)
        authService.loginResult = .success("token123")
        
        viewModel = LoginViewModel(
            authService: authService,
            networkMonitor: networkMonitor,
            tokenStorage: tokenStorage
        )
        
        viewModel.updateUsername("user")
        viewModel.updatePassword("password")
        try? await Task.sleep(nanoseconds: 100_000_000)
        
        // When: Valid credentials are entered and login is called
        viewModel.login()
        
        // Wait for async operation
        try? await Task.sleep(nanoseconds: 1_500_000_000) // 1.5 seconds
        
        // Then: Navigation event should be triggered
        XCTAssertNotNil(viewModel.state.navigationEvent, "Navigation event should be set")
        if case .navigateToHome = viewModel.state.navigationEvent {
            // Success
        } else {
            XCTFail("Navigation event should be navigateToHome")
        }
        XCTAssertNil(viewModel.state.errorMessage, "Error message should be null on success")
        XCTAssertEqual(viewModel.state.failureCount, 0, "Failure count should be reset")
    }
    
    func testErrorIncrementsFailureCount() async {
        // Given: Network is online and invalid credentials
        networkMonitor.isOnlineSubject.send(true)
        authService.loginResult = .failure(AuthError.invalidCredentials)
        
        viewModel = LoginViewModel(
            authService: authService,
            networkMonitor: networkMonitor,
            tokenStorage: tokenStorage
        )
        
        viewModel.updateUsername("wrong")
        viewModel.updatePassword("wrong")
        try? await Task.sleep(nanoseconds: 100_000_000)
        
        // When: Login fails
        viewModel.login()
        try? await Task.sleep(nanoseconds: 1_500_000_000)
        
        // Then: Failure count should be incremented
        XCTAssertEqual(viewModel.state.failureCount, 1, "Failure count should be 1")
        XCTAssertNotNil(viewModel.state.errorMessage, "Error message should be set")
        XCTAssertFalse(viewModel.state.isLockedOut, "Should not be locked out after 1 failure")
        
        // When: Login fails again
        viewModel.login()
        try? await Task.sleep(nanoseconds: 1_500_000_000)
        XCTAssertEqual(viewModel.state.failureCount, 2, "Failure count should be 2")
        XCTAssertFalse(viewModel.state.isLockedOut, "Should not be locked out after 2 failures")
    }
    
    func testLockoutAfter3Failures() async {
        // Given: Network is online and invalid credentials
        networkMonitor.isOnlineSubject.send(true)
        authService.loginResult = .failure(AuthError.invalidCredentials)
        
        viewModel = LoginViewModel(
            authService: authService,
            networkMonitor: networkMonitor,
            tokenStorage: tokenStorage
        )
        
        viewModel.updateUsername("wrong")
        viewModel.updatePassword("wrong")
        try? await Task.sleep(nanoseconds: 100_000_000)
        
        // When: Login fails 3 times
        viewModel.login()
        try? await Task.sleep(nanoseconds: 1_500_000_000)
        XCTAssertEqual(viewModel.state.failureCount, 1, "Failure count should be 1")
        
        viewModel.login()
        try? await Task.sleep(nanoseconds: 1_500_000_000)
        XCTAssertEqual(viewModel.state.failureCount, 2, "Failure count should be 2")
        
        viewModel.login()
        try? await Task.sleep(nanoseconds: 1_500_000_000)
        
        // Then: Account should be locked out
        XCTAssertEqual(viewModel.state.failureCount, 3, "Failure count should be 3")
        XCTAssertTrue(viewModel.state.isLockedOut, "Account should be locked out")
        XCTAssertFalse(viewModel.state.isButtonEnabled, "Button should be disabled when locked out")
    }
    
    func testOfflineShowsMessageAndNoServiceCall() async {
        // Given: Network is offline
        networkMonitor.isOnlineSubject.send(false)
        
        viewModel = LoginViewModel(
            authService: authService,
            networkMonitor: networkMonitor,
            tokenStorage: tokenStorage
        )
        
        viewModel.updateUsername("user")
        viewModel.updatePassword("password")
        try? await Task.sleep(nanoseconds: 100_000_000)
        
        // When: User tries to login while offline
        // Then: Offline message should be shown
        XCTAssertTrue(viewModel.state.isOffline, "Should show offline state")
        XCTAssertFalse(viewModel.state.isButtonEnabled, "Button should be disabled when offline")
        
        // When: Login is attempted (should not call service)
        viewModel.login()
        try? await Task.sleep(nanoseconds: 500_000_000)
        
        // Then: AuthService should not be called
        XCTAssertEqual(authService.loginCallCount, 0, "AuthService should not be called when offline")
    }
    
    func testRememberMePersistsToken() async {
        // Given: Network is online and valid credentials
        let testToken = "test_token_123"
        networkMonitor.isOnlineSubject.send(true)
        authService.loginResult = .success(testToken)
        
        viewModel = LoginViewModel(
            authService: authService,
            networkMonitor: networkMonitor,
            tokenStorage: tokenStorage
        )
        
        viewModel.updateUsername("user")
        viewModel.updatePassword("password")
        viewModel.updateRememberMe(true)
        try? await Task.sleep(nanoseconds: 100_000_000)
        
        // When: Login succeeds with remember me enabled
        viewModel.login()
        try? await Task.sleep(nanoseconds: 1_500_000_000)
        
        // Then: Token should be saved
        XCTAssertTrue(tokenStorage.saveTokenCalled, "Token should be saved")
        XCTAssertEqual(tokenStorage.savedToken, testToken, "Correct token should be saved")
        
        // When: Remember me is disabled
        viewModel.updateRememberMe(false)
        viewModel.updateUsername("user2")
        viewModel.updatePassword("password2")
        authService.loginResult = .success("token2")
        viewModel.login()
        try? await Task.sleep(nanoseconds: 1_500_000_000)
        
        // Then: Token should be cleared
        XCTAssertTrue(tokenStorage.clearTokenCalled, "Token should be cleared")
    }
}

// MARK: - Mock Classes

class MockAuthService: AuthServiceProtocol {
    var loginResult: Result<String, Error> = .success("default_token")
    var loginCallCount = 0
    
    func login(username: String, password: String) async throws -> String {
        loginCallCount += 1
        try await Task.sleep(nanoseconds: 100_000_000) // 0.1 seconds for testing
        switch loginResult {
        case .success(let token):
            return token
        case .failure(let error):
            throw error
        }
    }
}

class MockNetworkMonitor: NetworkMonitorProtocol {
    let isOnlineSubject = CurrentValueSubject<Bool, Never>(true)
    
    var isOnline: AnyPublisher<Bool, Never> {
        isOnlineSubject.eraseToAnyPublisher()
    }
}

class MockTokenStorage: TokenStorageProtocol {
    var savedToken: String?
    var saveTokenCalled = false
    var clearTokenCalled = false
    
    func saveToken(_ token: String) async {
        savedToken = token
        saveTokenCalled = true
    }
    
    func getToken() async -> String? {
        return savedToken
    }
    
    func clearToken() async {
        savedToken = nil
        clearTokenCalled = true
    }
}

